PHPCompatibility/Sniffs/Syntax/NewClassMemberAccessSniff.php

```diff
@@ -56,58 +61,124 @@ public function register()
      */
     public function process(File $phpcsFile, $stackPtr)
     {
-        $tokens = $phpcsFile->getTokens();
-
-        if ($tokens[$stackPtr]['code'] === \T_NEW && $this->supportsBelow('5.3') !== true) {
+        if ($this->supportsBelow('5.6') === false) {
             return;
-        } elseif ($tokens[$stackPtr]['code'] === \T_CLONE && $this->supportsBelow('5.6') !== true) {
+        }
+
+        $pointers = $this->isClassMemberAccess($phpcsFile, $stackPtr);
+        if (empty($pointers)) {
             return;
         }
 
+        $tokens     = $phpcsFile->getTokens();
+        $supports53 = $this->supportsBelow('5.3');
+
+        $error     = 'Class member access on object %s was not supported in PHP %s or earlier';
+        $data      = array('instantiation', '5.3');
+        $errorCode = 'OnNewFound';
+
+        if ($tokens[$stackPtr]['code'] === \T_CLONE) {
+            $data      = array('cloning', '5.6');
+            $errorCode = 'OnCloneFound';
+        }
+
+        foreach ($pointers as $open => $close) {
+            $itemData      = $data;
+            $itemErrorCode = $errorCode;
+
+            if ($tokens[$stackPtr]['code'] === \T_NEW
+                && $tokens[$open]['code'] !== \T_OPEN_CURLY_BRACKET
+            ) {
+                if ($supports53 === true) {
+                    $phpcsFile->addError($error, $open, $itemErrorCode, $itemData);
+                }
+                continue;
+            }
+
+            if ($tokens[$stackPtr]['code'] === \T_NEW
+                && $tokens[$open]['code'] === \T_OPEN_CURLY_BRACKET
+            ) {
+                // Non-curlies was already handled above.
+                $itemData      = array('instantiation using curly braces', '5.6');
+                $itemErrorCode = 'OnNewFoundUsingCurlies';
+            }
+
+            $phpcsFile->addError($error, $open, $itemErrorCode, $itemData);
+        }
+    }
+
+
+    /**
+     * Check if the class being instantiated/cloned is being dereferenced.
+     *
+     * @since 9.3.0 Logic split off from the process method.
+     *
+     * @param \PHP_CodeSniffer_File $phpcsFile The file being scanned.
+     * @param int                   $stackPtr  The position of the current token in
+     *                                         the stack passed in $tokens.
+     *
+     * @return array Array containing the stack pointers to the object operator or
+     *               the open/close braces involved in the class member access;
+     *               or an empty array if no class member access was detected.
+     */
+    public function isClassMemberAccess(File $phpcsFile, $stackPtr)
```

Comment:
Same comment as above : if this class returns an array, should we use "is*" as name ?